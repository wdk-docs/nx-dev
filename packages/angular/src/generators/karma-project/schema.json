{
  "$schema": "http://json-schema.org/schema",
  "$id": "NxKarmaProjectGenerator",
  "title": "Create Karma Configuration for a project",
  "cli": "nx",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "skipFormat": {
      "description": "Skip formatting files.",
      "type": "boolean",
      "default": false
    }
  },
  "additionalProperties": false,
  "required": ["project"]
}
