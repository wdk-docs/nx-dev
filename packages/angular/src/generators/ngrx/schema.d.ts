export interface NgRxGeneratorOptions {
  directory: string;
  minimal: boolean;
  module: string;
  name: string;
  useDataPersistence: boolean;
  barrels?: boolean;
  facade?: boolean;
  root?: boolean;
  skipFormat?: boolean;
  skipImport?: boolean;
  skipPackageJson?: boolean;
  syntax?: 'classes' | 'creators';
}
