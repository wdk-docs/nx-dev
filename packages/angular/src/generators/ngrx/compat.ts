import { convertNxGenerator } from '@nrwl/devkit';
import { ngrxGenerator } from './ngrx';

export default convertNxGenerator(ngrxGenerator);
