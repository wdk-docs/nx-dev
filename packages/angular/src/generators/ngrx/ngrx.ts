import type { GeneratorCallback, Tree } from '@nrwl/devkit';
import { formatFiles } from '@nrwl/devkit';
import {
  addExportsToBarrel,
  addImportsToModule,
  addNgRxToPackageJson,
  generateNgrxFilesFromTemplates,
  normalizeOptions,
} from './lib';
import type { NgRxGeneratorOptions } from './schema';

export async function ngrxGenerator(
  tree: Tree,
  options: NgRxGeneratorOptions
): Promise<GeneratorCallback> {
  const normalizedOptions = normalizeOptions(options);

  if (!tree.exists(normalizedOptions.module)) {
    throw new Error(`Module does not exist: ${normalizedOptions.module}.`);
  }

  if (!normalizedOptions.minimal || !normalizedOptions.root) {
    generateNgrxFilesFromTemplates(tree, normalizedOptions);
  }

  if (!normalizedOptions.skipImport) {
    addImportsToModule(tree, normalizedOptions);
    addExportsToBarrel(tree, normalizedOptions);
  }

  let packageInstallationTask: GeneratorCallback = () => {};
  if (!normalizedOptions.skipPackageJson) {
    packageInstallationTask = addNgRxToPackageJson(tree);
  }

  if (!normalizedOptions.skipFormat) {
    await formatFiles(tree);
  }

  return packageInstallationTask;
}

export default ngrxGenerator;
