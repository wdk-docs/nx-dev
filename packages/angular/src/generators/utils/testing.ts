import type { Tree } from '@nrwl/devkit';
import { wrapAngularDevkitSchematic } from '@nrwl/devkit/ngcli-adapter';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { Linter } from '@nrwl/linter';
import { UnitTestRunner } from '../../utils/test-runners';
import libraryGenerator from '../library/library';

export async function createStorybookTestWorkspaceForLib(
  libName: string
): Promise<Tree> {
  let tree = createTreeWithEmptyWorkspace();

  const moduleGenerator = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'module'
  );
  const componentGenerator = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'component'
  );

  await libraryGenerator(tree, {
    name: libName,
    buildable: false,
    linter: Linter.EsLint,
    publishable: false,
    simpleModuleName: false,
    skipFormat: false,
    unitTestRunner: UnitTestRunner.Jest,
  });

  await componentGenerator(tree, {
    name: 'test-button',
    project: libName,
  });

  tree.write(
    `libs/${libName}/src/lib/test-button/test-button.component.ts`,
    `import { Component, Input } from '@angular/core';

export type ButtonStyle = 'default' | 'primary' | 'accent';

@Component({
  selector: 'proj-test-button',
  templateUrl: './test-button.component.html',
  styleUrls: ['./test-button.component.css']
})
export class TestButtonComponent {
  @Input('buttonType') type = 'button';
  @Input() style: ButtonStyle = 'default';
  @Input() age?: number;
  @Input() isOn = false;
}`
  );

  tree.write(
    `libs/${libName}/src/lib/test-button/test-button.component.html`,
    `<button [attr.type]="type" [ngClass]="style"></button>`
  );

  const modulePath = `libs/${libName}/src/lib/${libName}.module.ts`;
  tree.write(
    modulePath,
    `import * as ButtonExports from './test-button/test-button.component';
    ${tree.read(modulePath)}`
  );

  // create a module with component that gets exported in a barrel file
  await moduleGenerator(tree, {
    name: 'barrel',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'barrel-button',
    project: libName,
    path: `libs/${libName}/src/lib/barrel`,
    module: 'barrel',
  });

  tree.write(
    `libs/${libName}/src/lib/barrel/barrel-button/index.ts`,
    `export * from './barrel-button.component';`
  );

  tree.write(
    `libs/${libName}/src/lib/barrel/barrel.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BarrelButtonComponent } from './barrel-button';

@NgModule({
  imports: [CommonModule],
  declarations: [BarrelButtonComponent],
})
export class BarrelModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  await moduleGenerator(tree, {
    name: 'variable-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-declare-button',
    project: libName,
    path: `libs/${libName}/src/lib/variable-declare`,
    module: 'variable-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-declare-view',
    project: libName,
    path: `libs/${libName}/src/lib/variable-declare`,
    module: 'variable-declare',
  });

  tree.write(
    `libs/${libName}/src/lib/variable-declare/variable-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableDeclareButtonComponent } from './variable-declare-button/variable-declare-button.component';
import { VariableDeclareViewComponent } from './variable-declare-view/variable-declare-view.component';

const COMPONENTS = [
  VariableDeclareButtonComponent,
  VariableDeclareViewComponent
]

@NgModule({
  imports: [CommonModule],
  declarations: COMPONENTS,
  exports: COMPONENTS
})
export class VariableDeclareModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  await moduleGenerator(tree, {
    name: 'variable-spread-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-button',
    project: libName,
    path: `libs/${libName}/src/lib/variable-spread-declare`,
    module: 'variable-spread-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-view',
    project: libName,
    path: `libs/${libName}/src/lib/variable-spread-declare`,
    module: 'variable-spread-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-anotherview',
    project: libName,
    path: `libs/${libName}/src/lib/variable-spread-declare`,
    module: 'variable-spread-declare',
  });

  tree.write(
    `libs/${libName}/src/lib/variable-spread-declare/variable-spread-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableSpreadDeclareButtonComponent } from './variable-spread-declare-button/variable-spread-declare-button.component';
import { VariableSpreadDeclareViewComponent } from './variable-spread-declare-view/variable-spread-declare-view.component';
import { VariableSpreadDeclareAnotherviewComponent } from './variable-spread-declare-anotherview/variable-spread-declare-anotherview.component';

const COMPONENTS = [ 
  VariableSpreadDeclareButtonComponent, 
  VariableSpreadDeclareViewComponent 
]

@NgModule({
  imports: [CommonModule],
  declarations: [...COMPONENTS, VariableSpreadDeclareAnotherviewComponent],
})
export class VariableSpreadDeclareModule {}`
  );

  // create a module where declared components are pulled from a static member of the module
  await moduleGenerator(tree, {
    name: 'static-member-declarations',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'cmp1',
    project: libName,
    path: `libs/${libName}/src/lib/static-member-declarations`,
    module: 'static-member-declarations',
  });

  await componentGenerator(tree, {
    name: 'cmp2',
    project: libName,
    path: `libs/${libName}/src/lib/static-member-declarations`,
    module: 'static-member-declarations',
  });

  tree.write(
    `libs/${libName}/src/lib/static-member-declarations/static-member-declarations.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Cmp1Component } from './cmp1/cmp1.component';
import { Cmp2Component } from './cmp2/cmp2.component';

@NgModule({
  imports: [CommonModule],
  declarations: StaticMemberDeclarationsModule.COMPONENTS,
  exports: StaticMemberDeclarationsModule.COMPONENTS
})
export class StaticMemberDeclarationsModule {
  static readonly COMPONENTS = [Cmp1Component, Cmp2Component];
}`
  );

  // create another button in a nested subpath
  await moduleGenerator(tree, {
    name: 'nested',
    project: libName,
    path: `libs/${libName}/src/lib`,
  });

  await componentGenerator(tree, {
    name: 'nested-button',
    project: libName,
    module: 'nested',
    path: `libs/${libName}/src/lib/nested`,
  });

  await componentGenerator(tree, {
    name: 'test-other',
    project: libName,
  });

  return tree;
}
