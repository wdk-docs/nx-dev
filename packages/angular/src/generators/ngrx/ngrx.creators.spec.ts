import { Tree } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { dirname } from 'path';
import {
  AppConfig,
  createApp,
  createLib,
  getAppConfig,
  getLibConfig,
} from '../../utils/nx-devkit/testing';
import { ngrxGenerator } from './ngrx';
import { NgRxGeneratorOptions } from './schema';

describe('NgRx generator', () => {
  let appConfig: AppConfig;
  let statePath: string;
  let tree: Tree;

  const defaultOptions: NgRxGeneratorOptions = {
    directory: '+state',
    minimal: true,
    module: 'apps/myapp/src/app/app.module.ts',
    name: 'users',
    useDataPersistence: false,
    syntax: 'creators',
  };

  const expectFileToExist = (file: string) =>
    expect(tree.exists(file)).toBeTruthy();
  const expectFileToNotExist = (file: string) =>
    expect(tree.exists(file)).not.toBeTruthy();

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    createApp(tree, 'myapp');
    appConfig = getAppConfig();
    statePath = `${dirname(appConfig.appModule)}/+state`;
  });

  describe('creators syntax', () => {
    it('should generate files without a facade', async () => {
      await ngrxGenerator(tree, {
        ...defaultOptions,
        module: appConfig.appModule,
      });

      expectFileToExist(`${statePath}/users.actions.ts`);
      expectFileToExist(`${statePath}/users.effects.ts`);
      expectFileToExist(`${statePath}/users.effects.spec.ts`);
      expectFileToExist(`${statePath}/users.models.ts`);
      expectFileToExist(`${statePath}/users.reducer.ts`);
      expectFileToExist(`${statePath}/users.reducer.spec.ts`);
      expectFileToExist(`${statePath}/users.selectors.ts`);
      expectFileToExist(`${statePath}/users.selectors.spec.ts`);
      expectFileToNotExist(`${statePath}/users.facade.ts`);
      expectFileToNotExist(`${statePath}/users.facade.spec.ts`);
    });

    it('should generate files with a facade', async () => {
      await ngrxGenerator(tree, {
        ...defaultOptions,
        module: appConfig.appModule,
        facade: true,
      });

      expectFileToExist(`${statePath}/users.actions.ts`);
      expectFileToExist(`${statePath}/users.effects.ts`);
      expectFileToExist(`${statePath}/users.effects.spec.ts`);
      expectFileToExist(`${statePath}/users.facade.ts`);
      expectFileToExist(`${statePath}/users.facade.spec.ts`);
      expectFileToExist(`${statePath}/users.models.ts`);
      expectFileToExist(`${statePath}/users.reducer.ts`);
      expectFileToExist(`${statePath}/users.reducer.spec.ts`);
      expectFileToExist(`${statePath}/users.selectors.ts`);
      expectFileToExist(`${statePath}/users.selectors.spec.ts`);
    });

    it('should generate the ngrx actions', async () => {
      await ngrxGenerator(tree, {
        ...defaultOptions,
        module: appConfig.appModule,
      });

      expect(
        tree.read(`${statePath}/users.actions.ts`, 'utf-8')
      ).toMatchSnapshot();
    });

    it('should generate the ngrx effects', async () => {
      await ngrxGenerator(tree, {
        ...defaultOptions,
        module: appConfig.appModule,
      });

      expect(
        tree.read(`${statePath}/users.effects.ts`, 'utf-8')
      ).toMatchSnapshot();
    });

    it('should generate the ngrx facade', async () => {
      await ngrxGenerator(tree, {
        ...defaultOptions,
        module: appConfig.appModule,
        facade: true,
      });

      expect(
        tree.read(`${statePath}/users.facade.ts`, 'utf-8')
      ).toMatchSnapshot();
    });

    it('should generate a models file for the feature', async () => {
      await ngrxGenerator(tree, {
        ...defaultOptions,
        module: appConfig.appModule,
        minimal: false,
      });

      expect(
        tree.read(`${statePath}/users.models.ts`, 'utf-8')
      ).toMatchSnapshot();
    });

    it('should generate the ngrx reducer', async () => {
      await ngrxGenerator(tree, {
        ...defaultOptions,
        module: appConfig.appModule,
      });

      expect(
        tree.read(`${statePath}/users.reducer.ts`, 'utf-8')
      ).toMatchSnapshot();
    });

    it('should generate the ngrx selectors', async () => {
      await ngrxGenerator(tree, {
        ...defaultOptions,
        module: appConfig.appModule,
      });

      expect(
        tree.read(`${statePath}/users.selectors.ts`, 'utf-8')
      ).toMatchSnapshot();
    });

    it('should use DataPersistence when useDataPersistence is true', async () => {
      await ngrxGenerator(tree, {
        ...defaultOptions,
        module: appConfig.appModule,
        minimal: false,
        useDataPersistence: true,
      });

      expect(
        tree.read(`${statePath}/users.effects.ts`, 'utf-8')
      ).toMatchSnapshot();
    });

    it('should generate with custom directory', async () => {
      statePath = '/apps/myapp/src/app/my-custom-directory';

      await ngrxGenerator(tree, {
        ...defaultOptions,
        directory: 'my-custom-directory',
        minimal: false,
        facade: true,
      });

      expectFileToExist(`${statePath}/users.actions.ts`);
      expectFileToExist(`${statePath}/users.effects.ts`);
      expectFileToExist(`${statePath}/users.effects.spec.ts`);
      expectFileToExist(`${statePath}/users.facade.ts`);
      expectFileToExist(`${statePath}/users.facade.spec.ts`);
      expectFileToExist(`${statePath}/users.models.ts`);
      expectFileToExist(`${statePath}/users.reducer.ts`);
      expectFileToExist(`${statePath}/users.reducer.spec.ts`);
      expectFileToExist(`${statePath}/users.selectors.ts`);
      expectFileToExist(`${statePath}/users.selectors.spec.ts`);
    });

    it('should update the entry point file with the right exports', async () => {
      createLib(tree, 'flights');
      let libConfig = getLibConfig();

      await ngrxGenerator(tree, {
        ...defaultOptions,
        name: 'super-users',
        module: libConfig.module,
        facade: true,
      });

      expect(tree.read(libConfig.barrel, 'utf-8')).toMatchSnapshot();
    });

    it('should update the entry point file correctly when barrels is true', async () => {
      createLib(tree, 'flights');
      let libConfig = getLibConfig();

      await ngrxGenerator(tree, {
        ...defaultOptions,
        name: 'super-users',
        module: libConfig.module,
        facade: true,
        barrels: true,
      });

      expect(tree.read(libConfig.barrel, 'utf-8')).toMatchSnapshot();
    });

    it('should update the entry point file with no facade', async () => {
      createLib(tree, 'flights');
      let libConfig = getLibConfig();

      await ngrxGenerator(tree, {
        ...defaultOptions,
        name: 'super-users',
        module: libConfig.module,
        facade: false,
      });

      expect(tree.read(libConfig.barrel, 'utf-8')).toMatchSnapshot();
    });

    describe('unit tests', () => {
      it('should generate specs for the ngrx effects', async () => {
        await ngrxGenerator(tree, {
          ...defaultOptions,
          name: 'super-users',
          module: appConfig.appModule,
          minimal: false,
        });

        expect(
          tree.read(`${statePath}/super-users.effects.spec.ts`, 'utf-8')
        ).toMatchSnapshot();
      });

      it('should generate specs for the ngrx effects correctly when useDataPersistence is true', async () => {
        await ngrxGenerator(tree, {
          ...defaultOptions,
          name: 'super-users',
          module: appConfig.appModule,
          minimal: false,
          useDataPersistence: true,
        });

        expect(
          tree.read(`${statePath}/super-users.effects.spec.ts`, 'utf-8')
        ).toMatchSnapshot();
      });

      it('should generate specs for the ngrx facade', async () => {
        await ngrxGenerator(tree, {
          ...defaultOptions,
          name: 'super-users',
          module: appConfig.appModule,
          minimal: false,
          facade: true,
        });

        expect(
          tree.read(`${statePath}/super-users.facade.spec.ts`, 'utf-8')
        ).toMatchSnapshot();
      });

      it('should generate specs for the ngrx reducer', async () => {
        await ngrxGenerator(tree, {
          ...defaultOptions,
          name: 'super-users',
          module: appConfig.appModule,
          minimal: false,
        });

        expect(
          tree.read(`${statePath}/super-users.reducer.spec.ts`, 'utf-8')
        ).toMatchSnapshot();
      });

      it('should generate specs for the ngrx selectors', async () => {
        await ngrxGenerator(tree, {
          ...defaultOptions,
          name: 'super-users',
          module: appConfig.appModule,
          minimal: false,
        });

        expect(
          tree.read(`${statePath}/super-users.selectors.spec.ts`, 'utf-8')
        ).toMatchSnapshot();
      });
    });
  });
});
