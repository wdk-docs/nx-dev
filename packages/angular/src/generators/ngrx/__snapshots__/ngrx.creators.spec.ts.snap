// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NgRx generator creators syntax should generate a models file for the feature 1`] = `
"/**
 * Interface for the 'Users' data
 */
export interface UsersEntity {
  id: string | number; // Primary ID
  name: string;
};"
`;

exports[`NgRx generator creators syntax should generate the ngrx actions 1`] = `
"import { createAction, props } from '@ngrx/store';
import { UsersEntity } from './users.models';

export const init = createAction(
  '[Users Page] Init'
);

export const loadUsersSuccess = createAction(
  '[Users/API] Load Users Success',
  props<{ users: UsersEntity[] }>()
);

export const loadUsersFailure = createAction(
  '[Users/API] Load Users Failure',
  props<{ error: any }>()
);
"
`;

exports[`NgRx generator creators syntax should generate the ngrx effects 1`] = `
"import { Injectable } from '@angular/core';
import { createEffect, Actions, ofType } from '@ngrx/effects';
import { fetch } from '@nrwl/angular';

import * as UsersActions from './users.actions';
import * as UsersFeature from './users.reducer';

@Injectable()
export class UsersEffects {
  init$ = createEffect(() => this.actions$.pipe(
    ofType(UsersActions.init),
    fetch({
      run: action => {
        // Your custom service 'load' logic goes here. For now just return a success action...
        return UsersActions.loadUsersSuccess({ users: [] });
      },
      onError: (action, error) => {
        console.error('Error', error);
        return UsersActions.loadUsersFailure({ error });
      }
    })
  ));

  constructor(
    private readonly actions$: Actions
  ) {}
}
"
`;

exports[`NgRx generator creators syntax should generate the ngrx facade 1`] = `
"import { Injectable } from '@angular/core';
import { select, Store, Action } from '@ngrx/store';

import * as UsersActions from './users.actions';
import * as UsersFeature from './users.reducer';
import * as UsersSelectors from './users.selectors';

@Injectable()
export class UsersFacade {
  /**
   * Combine pieces of state using createSelector, 
   * and expose them as observables through the facade.
   */
  loaded$ = this.store.pipe(select(UsersSelectors.getUsersLoaded));
  allUsers$ = this.store.pipe(select(UsersSelectors.getAllUsers));
  selectedUsers$ = this.store.pipe(select(UsersSelectors.getSelected));

  constructor(private readonly store: Store) {}

  /**
   * Use the initialization action to perform one 
   * or more tasks in your Effects.
   */
  init() {
    this.store.dispatch(UsersActions.init());
  }
}
"
`;

exports[`NgRx generator creators syntax should generate the ngrx reducer 1`] = `
"import { EntityState, EntityAdapter, createEntityAdapter } from '@ngrx/entity';
import { createReducer, on, Action } from '@ngrx/store';

import * as UsersActions from './users.actions';
import { UsersEntity } from './users.models';

export const USERS_FEATURE_KEY = 'users';

export interface State extends EntityState<UsersEntity> {
  selectedId?: string | number; // which Users record has been selected
  loaded: boolean; // has the Users list been loaded
  error?: string | null; // last known error (if any)
}

export interface UsersPartialState {
  readonly [USERS_FEATURE_KEY]: State;
}

export const usersAdapter: EntityAdapter<UsersEntity> = createEntityAdapter<UsersEntity>();

export const initialState: State = usersAdapter.getInitialState({
  // set initial required properties
  loaded: false
});

const usersReducer = createReducer(
  initialState,
  on(UsersActions.init,
    state => ({ ...state, loaded: false, error: null })
  ),
  on(UsersActions.loadUsersSuccess,
    (state, { users }) => usersAdapter.setAll(users, { ...state, loaded: true })
  ),
  on(UsersActions.loadUsersFailure,
    (state, { error }) => ({ ...state, error })
  ),
);

export function reducer(state: State | undefined, action: Action) {
  return usersReducer(state, action);
}
"
`;

exports[`NgRx generator creators syntax should generate the ngrx selectors 1`] = `
"import { createFeatureSelector, createSelector } from '@ngrx/store';
import { USERS_FEATURE_KEY, State, usersAdapter } from './users.reducer';

// Lookup the 'Users' feature state managed by NgRx
export const getUsersState = createFeatureSelector<State>(USERS_FEATURE_KEY);

const { selectAll, selectEntities } = usersAdapter.getSelectors();

export const getUsersLoaded = createSelector(
  getUsersState,
  (state: State) => state.loaded
);

export const getUsersError = createSelector(
  getUsersState,
  (state: State) => state.error
);

export const getAllUsers = createSelector(
  getUsersState,
  (state: State) => selectAll(state)
);

export const getUsersEntities = createSelector(
  getUsersState,
  (state: State) => selectEntities(state)
);

export const getSelectedId = createSelector(
  getUsersState,
  (state: State) => state.selectedId
);

export const getSelected = createSelector(
  getUsersEntities,
  getSelectedId,
  (entities, selectedId) => (selectedId ? entities[selectedId] : undefined)
);
"
`;

exports[`NgRx generator creators syntax should update the entry point file correctly when barrels is true 1`] = `
"import * as SuperUsersActions from './lib/+state/super-users.actions';

import * as SuperUsersFeature from './lib/+state/super-users.reducer';

import * as SuperUsersSelectors from './lib/+state/super-users.selectors';

export * from './lib/+state/super-users.facade';

export * from './lib/+state/super-users.models';

export { SuperUsersActions, SuperUsersFeature, SuperUsersSelectors };

    export * from './lib/flights.module';
  "
`;

exports[`NgRx generator creators syntax should update the entry point file with no facade 1`] = `
"export * from './lib/+state/super-users.models';
export * from './lib/+state/super-users.selectors';
export * from './lib/+state/super-users.reducer';
export * from './lib/+state/super-users.actions';

    export * from './lib/flights.module';
  "
`;

exports[`NgRx generator creators syntax should update the entry point file with the right exports 1`] = `
"export * from './lib/+state/super-users.facade';
export * from './lib/+state/super-users.models';
export * from './lib/+state/super-users.selectors';
export * from './lib/+state/super-users.reducer';
export * from './lib/+state/super-users.actions';

    export * from './lib/flights.module';
  "
`;

exports[`NgRx generator creators syntax should use DataPersistence when useDataPersistence is true 1`] = `
"import { Injectable } from '@angular/core';
import { createEffect, Actions, ofType } from '@ngrx/effects';
import { DataPersistence } from '@nrwl/angular';

import * as UsersActions from './users.actions';
import * as UsersFeature from './users.reducer';

@Injectable()
export class UsersEffects {
  init$ = createEffect(() => this.dataPersistence.fetch(UsersActions.init, {
    run: (action: ReturnType<typeof UsersActions.init>, state: UsersFeature.UsersPartialState) => {
      // Your custom service 'load' logic goes here. For now just return a success action...
      return UsersActions.loadUsersSuccess({ users: [] });
    },
    onError: (action: ReturnType<typeof UsersActions.init>, error) => {
      console.error('Error', error);
      return UsersActions.loadUsersFailure({ error });
    }
  }));

  constructor(
    private readonly actions$: Actions,
    private readonly dataPersistence: DataPersistence<UsersFeature.UsersPartialState>
  ) {}
}
"
`;

exports[`NgRx generator creators syntax unit tests should generate specs for the ngrx effects 1`] = `
"import { TestBed } from '@angular/core/testing';
import { provideMockActions } from '@ngrx/effects/testing';
import { Action } from '@ngrx/store';
import { provideMockStore } from '@ngrx/store/testing';
import { NxModule } from '@nrwl/angular';
import { hot } from 'jasmine-marbles';
import { Observable } from 'rxjs';

import * as SuperUsersActions from './super-users.actions';
import { SuperUsersEffects } from './super-users.effects';

describe('SuperUsersEffects', () => {
  let actions: Observable<Action>;
  let effects: SuperUsersEffects;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        NxModule.forRoot(),
      ],
      providers: [
        SuperUsersEffects,
        provideMockActions(() => actions),
        provideMockStore()
      ],
    });

    effects = TestBed.inject(SuperUsersEffects);
  });

  describe('init$', () => {
    it('should work', () => {
      actions = hot('-a-|', { a: SuperUsersActions.init() });

      const expected = hot('-a-|', { a: SuperUsersActions.loadSuperUsersSuccess({ superUsers: [] }) });

      expect(effects.init$).toBeObservable(expected);
    });
  });
});
"
`;

exports[`NgRx generator creators syntax unit tests should generate specs for the ngrx effects correctly when useDataPersistence is true 1`] = `
"import { TestBed } from '@angular/core/testing';
import { provideMockActions } from '@ngrx/effects/testing';
import { Action } from '@ngrx/store';
import { provideMockStore } from '@ngrx/store/testing';
import { NxModule, DataPersistence } from '@nrwl/angular';
import { hot } from 'jasmine-marbles';
import { Observable } from 'rxjs';

import * as SuperUsersActions from './super-users.actions';
import { SuperUsersEffects } from './super-users.effects';

describe('SuperUsersEffects', () => {
  let actions: Observable<Action>;
  let effects: SuperUsersEffects;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        NxModule.forRoot(),
      ],
      providers: [
        SuperUsersEffects,
        DataPersistence,
        provideMockActions(() => actions),
        provideMockStore()
      ],
    });

    effects = TestBed.inject(SuperUsersEffects);
  });

  describe('init$', () => {
    it('should work', () => {
      actions = hot('-a-|', { a: SuperUsersActions.init() });

      const expected = hot('-a-|', { a: SuperUsersActions.loadSuperUsersSuccess({ superUsers: [] }) });

      expect(effects.init$).toBeObservable(expected);
    });
  });
});
"
`;

exports[`NgRx generator creators syntax unit tests should generate specs for the ngrx facade 1`] = `
"import { NgModule } from '@angular/core';
import { TestBed } from '@angular/core/testing';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule, Store } from '@ngrx/store';
import { NxModule } from '@nrwl/angular';
import { readFirst } from '@nrwl/angular/testing';

import * as SuperUsersActions from './super-users.actions';
import { SuperUsersEffects } from './super-users.effects';
import { SuperUsersFacade } from './super-users.facade';
import { SuperUsersEntity } from './super-users.models';
import {
  SUPER_USERS_FEATURE_KEY,
  State,
  initialState,
  reducer
} from './super-users.reducer';
import * as SuperUsersSelectors from './super-users.selectors';

interface TestSchema {
  superUsers: State;
}

describe('SuperUsersFacade', () => {
  let facade: SuperUsersFacade;
  let store: Store<TestSchema>;
  const createSuperUsersEntity = (id: string, name = ''): SuperUsersEntity => ({
    id,
    name: name || \`name-\${id}\`
  });

  describe('used in NgModule', () => {
    beforeEach(() => {
      @NgModule({
        imports: [
          StoreModule.forFeature(SUPER_USERS_FEATURE_KEY, reducer),
          EffectsModule.forFeature([SuperUsersEffects])
        ],
        providers: [SuperUsersFacade]
      })
      class CustomFeatureModule {}

      @NgModule({
        imports: [
          NxModule.forRoot(),
          StoreModule.forRoot({}),
          EffectsModule.forRoot([]),
          CustomFeatureModule,
        ]
      })
      class RootModule {}
      TestBed.configureTestingModule({ imports: [RootModule] });

      store = TestBed.inject(Store);
      facade = TestBed.inject(SuperUsersFacade);
    });

    /**
     * The initially generated facade::loadAll() returns empty array
     */
    it('loadAll() should return empty list with loaded == true', async () => {
      let list = await readFirst(facade.allSuperUsers$);
      let isLoaded = await readFirst(facade.loaded$);

      expect(list.length).toBe(0);
      expect(isLoaded).toBe(false);

      facade.init();

      list = await readFirst(facade.allSuperUsers$);
      isLoaded = await readFirst(facade.loaded$);

      expect(list.length).toBe(0);
      expect(isLoaded).toBe(true);
    });

    /**
     * Use \`loadSuperUsersSuccess\` to manually update list
     */
    it('allSuperUsers$ should return the loaded list; and loaded flag == true', async () => {
      let list = await readFirst(facade.allSuperUsers$);
      let isLoaded = await readFirst(facade.loaded$);

      expect(list.length).toBe(0);
      expect(isLoaded).toBe(false);

      store.dispatch(SuperUsersActions.loadSuperUsersSuccess({
        superUsers: [
          createSuperUsersEntity('AAA'),
          createSuperUsersEntity('BBB')
        ]})
      );

      list = await readFirst(facade.allSuperUsers$);
      isLoaded = await readFirst(facade.loaded$);

      expect(list.length).toBe(2);
      expect(isLoaded).toBe(true);
    });
  });
});
"
`;

exports[`NgRx generator creators syntax unit tests should generate specs for the ngrx reducer 1`] = `
"import { Action } from '@ngrx/store';

import * as SuperUsersActions from './super-users.actions';
import { SuperUsersEntity } from './super-users.models';
import { State, initialState, reducer } from './super-users.reducer';

describe('SuperUsers Reducer', () => {
  const createSuperUsersEntity = (id: string, name = ''): SuperUsersEntity => ({
    id,
    name: name || \`name-\${id}\`
  });

  describe('valid SuperUsers actions', () => {
    it('loadSuperUsersSuccess should return the list of known SuperUsers', () => {
      const superUsers = [
        createSuperUsersEntity('PRODUCT-AAA'),
        createSuperUsersEntity('PRODUCT-zzz')
      ];
      const action = SuperUsersActions.loadSuperUsersSuccess({ superUsers });

      const result: State = reducer(initialState, action);

      expect(result.loaded).toBe(true);
      expect(result.ids.length).toBe(2);
    });
  });

  describe('unknown action', () => {
    it('should return the previous state', () => {
      const action = {} as Action;

      const result = reducer(initialState, action);

      expect(result).toBe(initialState);
    });
  });
});
"
`;

exports[`NgRx generator creators syntax unit tests should generate specs for the ngrx selectors 1`] = `
"import { SuperUsersEntity } from './super-users.models';
import { superUsersAdapter, SuperUsersPartialState, initialState } from './super-users.reducer';
import * as SuperUsersSelectors from './super-users.selectors';

describe('SuperUsers Selectors', () => {
  const ERROR_MSG = 'No Error Available';
  const getSuperUsersId = (it: SuperUsersEntity) => it.id;
  const createSuperUsersEntity = (id: string, name = '') => ({
    id,
    name: name || \`name-\${id}\`
  }) as SuperUsersEntity;

  let state: SuperUsersPartialState;

  beforeEach(() => {
    state = {
      superUsers: superUsersAdapter.setAll([
        createSuperUsersEntity('PRODUCT-AAA'),
        createSuperUsersEntity('PRODUCT-BBB'),
        createSuperUsersEntity('PRODUCT-CCC')
      ], {
        ...initialState,
        selectedId : 'PRODUCT-BBB',
        error: ERROR_MSG,
        loaded: true
      })
    };
  });

  describe('SuperUsers Selectors', () => {
    it('getAllSuperUsers() should return the list of SuperUsers', () => {
      const results = SuperUsersSelectors.getAllSuperUsers(state);
      const selId = getSuperUsersId(results[1]);

      expect(results.length).toBe(3);
      expect(selId).toBe('PRODUCT-BBB');
    });

    it('getSelected() should return the selected Entity', () => {
      const result = SuperUsersSelectors.getSelected(state) as SuperUsersEntity;
      const selId = getSuperUsersId(result);

      expect(selId).toBe('PRODUCT-BBB');
    });

    it('getSuperUsersLoaded() should return the current \\"loaded\\" status', () => {
      const result = SuperUsersSelectors.getSuperUsersLoaded(state);

      expect(result).toBe(true);
    });

    it('getSuperUsersError() should return the current \\"error\\" state', () => {
      const result = SuperUsersSelectors.getSuperUsersError(state);

      expect(result).toBe(ERROR_MSG);
    });
  });
});
"
`;
